File: shared-swift/Sources/DirigibleCore/LocalCache.swift
import Foundation
import SQLite

/// LocalCache - SQLite-based local cache for instant reads
///
/// Architecture:
/// - Writes: SwiftUI → Firestore (direct) → LocalCache (mirror)
/// - Reads: SwiftUI → LocalCache (instant)
/// - Sync: Firestore listener → LocalCache (background)
public actor LocalCache {
    private var db: Connection?
    private let dbPath: URL

    // Tables
    private let notes = Table("notes")
    private let songs = Table("songs")

    // Note columns
    private let noteId = Expression<String>("id")
    private let noteType = Expression<String>("type")
    private let noteTitle = Expression<String>("title")
    private let noteParentId = Expression<String?>("parentId")
    private let noteContent = Expression<String?>("content")
    private let noteDate = Expression<String?>("date")
    private let noteTime = Expression<String?>("time")
    private let noteTags = Expression<String?>("tags")
    private let noteImages = Expression<String?>("images")
    private let noteEmbeddedMedia = Expression<String?>("embeddedMedia")
    private let noteGridSize = Expression<String?>("gridSize")
    private let noteSortMode = Expression<String?>("sortMode")
    private let noteMusicSortColumn = Expression<String?>("musicSortColumn")
    private let noteMusicSortDirection = Expression<String?>("musicSortDirection")
    private let notePublished = Expression<Bool>("published")
    private let noteSlug = Expression<String?>("slug")
    private let noteSortOrder = Expression<Int?>("sortOrder")
    private let noteCreatedAt = Expression<Double>("createdAt")
    private let noteUpdatedAt = Expression<Double>("updatedAt")

    // Song columns
    private let songId = Expression<String>("id")
    private let songLibraryId = Expression<String>("libraryId")
    private let songTitle = Expression<String>("title")
    private let songArtist = Expression<String>("artist")
    private let songAlbum = Expression<String>("album")
    private let songYear = Expression<Int?>("year")
    private let songTrackNumber = Expression<Int?>("trackNumber")
    private let songDiscNumber = Expression<Int?>("discNumber")
    private let songOriginalTrackNumber = Expression<Int?>("originalTrackNumber")
    private let songGenre = Expression<String>("genre")
    private let songDuration = Expression<Double>("duration")
    private let songFileSize = Expression<Int>("fileSize")
    private let songAlbumArtUrl = Expression<String?>("albumArtUrl")
    private let songAlbumArtThumbUrl = Expression<String?>("albumArtThumbUrl")
    private let songStorageUrl = Expression<String>("storageUrl")
    private let songStoragePath = Expression<String>("storagePath")
    private let songFileName = Expression<String>("fileName")
    private let songFileType = Expression<String>("fileType")
    private let songDateAdded = Expression<Double>("dateAdded")
    private let songCreatedAt = Expression<Double>("createdAt")
    private let songUpdatedAt = Expression<Double>("updatedAt")

    // Singleton
    public static let shared = LocalCache()

    private init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dirigibleDir = appSupport.appendingPathComponent("Dirigible", isDirectory: true)

        // Create directory if needed
        try? FileManager.default.createDirectory(at: dirigibleDir, withIntermediateDirectories: true)

        self.dbPath = dirigibleDir.appendingPathComponent("cache.sqlite")
    }

    /// Initialize the database
    public func initialize() throws {
        db = try Connection(dbPath.path)

        // Create notes table
        try db?.run(notes.create(ifNotExists: true) { t in
            t.column(noteId, primaryKey: true)
            t.column(noteType)
            t.column(noteTitle)
            t.column(noteParentId)
            t.column(noteContent)
            t.column(noteDate)
            t.column(noteTime)
            t.column(noteTags)
            t.column(noteImages)
            t.column(noteEmbeddedMedia)
            t.column(noteGridSize)
            t.column(noteSortMode)
            t.column(noteMusicSortColumn)
            t.column(noteMusicSortDirection)
            t.column(notePublished)
            t.column(noteSlug)
            t.column(noteSortOrder)
            t.column(noteCreatedAt)
            t.column(noteUpdatedAt)
        })

        // Create index for parent queries
        try db?.run(notes.createIndex(noteParentId, ifNotExists: true))

        // Create songs table
        try db?.run(songs.create(ifNotExists: true) { t in
            t.column(songId, primaryKey: true)
            t.column(songLibraryId)
            t.column(songTitle)
            t.column(songArtist)
            t.column(songAlbum)
            t.column(songYear)
            t.column(songTrackNumber)
            t.column(songDiscNumber)
            t.column(songOriginalTrackNumber)
            t.column(songGenre)
            t.column(songDuration)
            t.column(songFileSize)
            t.column(songAlbumArtUrl)
            t.column(songAlbumArtThumbUrl)
            t.column(songStorageUrl)
            t.column(songStoragePath)
            t.column(songFileName)
            t.column(songFileType)
            t.column(songDateAdded)
            t.column(songCreatedAt)
            t.column(songUpdatedAt)
        })

        // Create index for library queries
        try db?.run(songs.createIndex(songLibraryId, ifNotExists: true))

        print("[LocalCache] SQLite initialized at \(dbPath.path)")
    }

    // MARK: - Notes

    /// Get all notes
    public func getAllNotes() throws -> [NoteItem] {
        guard let db else { throw CacheError.notInitialized }

        let query = notes.order(noteSortOrder.asc, noteCreatedAt.asc)
        return try db.prepare(query).map { row in
            try rowToNote(row)
        }
    }

    /// Get notes by parent ID
    public func getNotesByParent(_ parentId: String?) throws -> [NoteItem] {
        guard let db else { throw CacheError.notInitialized }

        let query: Table
        if let parentId {
            query = notes.filter(noteParentId == parentId).order(noteSortOrder.asc, noteCreatedAt.asc)
        } else {
            query = notes.filter(noteParentId == nil).order(noteSortOrder.asc, noteCreatedAt.asc)
        }

        return try db.prepare(query).map { row in
            try rowToNote(row)
        }
    }

    /// Get a single note by ID
    public func getNoteById(_ id: String) throws -> NoteItem? {
        guard let db else { throw CacheError.notInitialized }

        let query = notes.filter(noteId == id)
        guard let row = try db.pluck(query) else { return nil }
        return try rowToNote(row)
    }

    /// Upsert a note
    public func upsertNote(_ note: NoteItem) throws {
        guard let db else { throw CacheError.notInitialized }

        let encoder = JSONEncoder()
        let tagsJson = note.tags.map { try? encoder.encode($0) }.flatMap { $0.map { String(data: $0, encoding: .utf8) } } ?? nil
        let imagesJson = note.images.map { try? encoder.encode($0) }.flatMap { $0.map { String(data: $0, encoding: .utf8) } } ?? nil
        let embeddedMediaJson = note.embeddedMedia.map { try? encoder.encode($0) }.flatMap { $0.map { String(data: $0, encoding: .utf8) } } ?? nil

        try db.run(notes.upsert(
            noteId <- note.id,
            noteType <- note.type.rawValue,
            noteTitle <- note.title,
            noteParentId <- note.parentId,
            noteContent <- note.content,
            noteDate <- note.date,
            noteTime <- note.time,
            noteTags <- tagsJson,
            noteImages <- imagesJson,
            noteEmbeddedMedia <- embeddedMediaJson,
            noteGridSize <- note.gridSize?.rawValue,
            noteSortMode <- note.sortMode?.rawValue,
            noteMusicSortColumn <- note.musicSortColumn?.rawValue,
            noteMusicSortDirection <- note.musicSortDirection?.rawValue,
            notePublished <- (note.published ?? false),
            noteSlug <- note.slug,
            noteSortOrder <- note.sortOrder,
            noteCreatedAt <- note.createdAt.timeIntervalSince1970 * 1000,
            noteUpdatedAt <- note.updatedAt.timeIntervalSince1970 * 1000,
            onConflictOf: noteId
        ))
    }

    /// Bulk upsert notes
    public func upsertNotes(_ notes: [NoteItem]) throws {
        for note in notes {
            try upsertNote(note)
        }
    }

    /// Delete a note
    public func deleteNote(_ id: String) throws {
        guard let db else { throw CacheError.notInitialized }
        try db.run(notes.filter(noteId == id).delete())
    }

    private func rowToNote(_ row: Row) throws -> NoteItem {
        let decoder = JSONDecoder()

        var tags: [String]?
        if let tagsJson = row[noteTags], let data = tagsJson.data(using: .utf8) {
            tags = try? decoder.decode([String].self, from: data)
        }

        var images: [MoodboardImage]?
        if let imagesJson = row[noteImages], let data = imagesJson.data(using: .utf8) {
            images = try? decoder.decode([MoodboardImage].self, from: data)
        }

        var embeddedMedia: [EmbeddedMedia]?
        if let mediaJson = row[noteEmbeddedMedia], let data = mediaJson.data(using: .utf8) {
            embeddedMedia = try? decoder.decode([EmbeddedMedia].self, from: data)
        }

        return NoteItem(
            id: row[noteId],
            type: NoteType(rawValue: row[noteType]) ?? .note,
            title: row[noteTitle],
            parentId: row[noteParentId],
            createdAt: Date(timeIntervalSince1970: row[noteCreatedAt] / 1000),
            updatedAt: Date(timeIntervalSince1970: row[noteUpdatedAt] / 1000),
            content: row[noteContent],
            date: row[noteDate],
            time: row[noteTime],
            tags: tags,
            embeddedMedia: embeddedMedia,
            images: images,
            gridSize: row[noteGridSize].flatMap { GridSize(rawValue: $0) },
            sortMode: row[noteSortMode].flatMap { SortMode(rawValue: $0) },
            musicSortColumn: row[noteMusicSortColumn].flatMap { MusicSortColumn(rawValue: $0) },
            musicSortDirection: row[noteMusicSortDirection].flatMap { SortDirection(rawValue: $0) },
            published: row[notePublished],
            slug: row[noteSlug],
            sortOrder: row[noteSortOrder]
        )
    }

    // MARK: - Songs

    /// Get songs by library ID
    public func getSongsByLibrary(_ libraryId: String) throws -> [Song] {
        guard let db else { throw CacheError.notInitialized }

        let query = songs.filter(songLibraryId == libraryId)
        return try db.prepare(query).map { row in
            rowToSong(row)
        }
    }

    /// Get a single song by ID
    public func getSongById(_ id: String) throws -> Song? {
        guard let db else { throw CacheError.notInitialized }

        let query = songs.filter(songId == id)
        guard let row = try db.pluck(query) else { return nil }
        return rowToSong(row)
    }

    /// Upsert a song
    public func upsertSong(_ song: Song) throws {
        guard let db else { throw CacheError.notInitialized }

        try db.run(songs.upsert(
            songId <- song.id,
            songLibraryId <- song.libraryId,
            songTitle <- song.title,
            songArtist <- song.artist,
            songAlbum <- song.album,
            songYear <- song.year,
            songTrackNumber <- song.trackNumber,
            songDiscNumber <- song.discNumber,
            songOriginalTrackNumber <- song.originalTrackNumber,
            songGenre <- song.genre,
            songDuration <- song.duration,
            songFileSize <- song.fileSize,
            songAlbumArtUrl <- song.albumArtUrl,
            songAlbumArtThumbUrl <- song.albumArtThumbUrl,
            songStorageUrl <- song.storageUrl,
            songStoragePath <- song.storagePath,
            songFileName <- song.fileName,
            songFileType <- song.fileType,
            songDateAdded <- song.dateAdded.timeIntervalSince1970 * 1000,
            songCreatedAt <- song.createdAt.timeIntervalSince1970 * 1000,
            songUpdatedAt <- song.updatedAt.timeIntervalSince1970 * 1000,
            onConflictOf: songId
        ))
    }

    /// Bulk upsert songs
    public func upsertSongs(_ songs: [Song]) throws {
        for song in songs {
            try upsertSong(song)
        }
    }

    /// Delete a song
    public func deleteSong(_ id: String) throws {
        guard let db else { throw CacheError.notInitialized }
        try db.run(songs.filter(songId == id).delete())
    }

    /// Delete all songs in a library
    public func deleteSongsByLibrary(_ libraryId: String) throws {
        guard let db else { throw CacheError.notInitialized }
        try db.run(songs.filter(songLibraryId == libraryId).delete())
    }

    private func rowToSong(_ row: Row) -> Song {
        Song(
            id: row[songId],
            libraryId: row[songLibraryId],
            title: row[songTitle],
            artist: row[songArtist],
            album: row[songAlbum],
            year: row[songYear],
            trackNumber: row[songTrackNumber],
            discNumber: row[songDiscNumber],
            originalTrackNumber: row[songOriginalTrackNumber],
            genre: row[songGenre],
            duration: row[songDuration],
            fileSize: row[songFileSize],
            albumArtUrl: row[songAlbumArtUrl],
            albumArtThumbUrl: row[songAlbumArtThumbUrl],
            storageUrl: row[songStorageUrl],
            storagePath: row[songStoragePath],
            fileName: row[songFileName],
            fileType: row[songFileType],
            dateAdded: Date(timeIntervalSince1970: row[songDateAdded] / 1000),
            createdAt: Date(timeIntervalSince1970: row[songCreatedAt] / 1000),
            updatedAt: Date(timeIntervalSince1970: row[songUpdatedAt] / 1000)
        )
    }

    // MARK: - Utilities

    /// Clear all cached data
    public func clear() throws {
        guard let db else { throw CacheError.notInitialized }
        try db.run(notes.delete())
        try db.run(songs.delete())
        print("[LocalCache] Cleared all data")
    }

    /// Get cache statistics
    public func getStats() throws -> (notes: Int, songs: Int) {
        guard let db else { throw CacheError.notInitialized }
        let notesCount = try db.scalar(notes.count)
        let songsCount = try db.scalar(songs.count)
        return (notesCount, songsCount)
    }
}

// MARK: - Errors

public enum CacheError: Error, LocalizedError {
    case notInitialized

    public var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "Local cache has not been initialized"
        }
    }
}

File: src/lib/local-cache.ts
"use client";

import type { Database } from "sql.js";
import { Song } from "./songs";
import { NoteItem, MoodboardImage } from "./notes";
import { Timestamp } from "firebase/firestore";

// Dynamic import of sql.js to avoid SSR issues
async function loadSqlJs() {
  const initSqlJs = (await import("sql.js")).default;
  return initSqlJs;
}

// Singleton instance
let cacheInstance: LocalCache | null = null;
let initPromise: Promise<LocalCache> | null = null;

/**
 * LocalCache - SQLite-based local cache for instant reads
 *
 * Architecture:
 * - Writes: React → Firestore (direct) → LocalCache (mirror)
 * - Reads: React → LocalCache (instant)
 * - Sync: Firestore onSnapshot → LocalCache (background)
 */
export class LocalCache {
  private db: Database;
  private initialized = false;

  private constructor(db: Database) {
    this.db = db;
  }

  /**
   * Initialize the cache - call once at app startup
   */
  static async init(): Promise<LocalCache> {
    if (cacheInstance) return cacheInstance;
    if (initPromise) return initPromise;

    initPromise = (async () => {
      console.log("[LocalCache] Initializing SQLite...");

      // Load sql.js WASM dynamically
      const initSqlJs = await loadSqlJs();
      const SQL = await initSqlJs({
        locateFile: (file: string) => `https://sql.js.org/dist/${file}`,
      });

      // Try to load persisted database from IndexedDB
      let data: Uint8Array | null = null;
      try {
        data = await loadFromIndexedDB();
        if (data) {
          console.log("[LocalCache] Loaded persisted database from IndexedDB");
        }
      } catch (e) {
        console.warn("[LocalCache] Failed to load from IndexedDB:", e);
      }

      // Create database (from persisted data or fresh)
      const db = data ? new SQL.Database(data) : new SQL.Database();

      cacheInstance = new LocalCache(db);
      cacheInstance.createTables();
      cacheInstance.initialized = true;

      console.log("[LocalCache] SQLite ready");
      return cacheInstance;
    })();

    return initPromise;
  }

  /**
   * Get the singleton instance (throws if not initialized)
   */
  static getInstance(): LocalCache {
    if (!cacheInstance) {
      throw new Error("LocalCache not initialized. Call LocalCache.init() first.");
    }
    return cacheInstance;
  }

  /**
   * Check if cache is initialized
   */
  static isInitialized(): boolean {
    return cacheInstance !== null && cacheInstance.initialized;
  }

  /**
   * Create database tables
   */
  private createTables() {
    // Songs table
    this.db.run(`
      CREATE TABLE IF NOT EXISTS songs (
        id TEXT PRIMARY KEY,
        libraryId TEXT NOT NULL,
        title TEXT NOT NULL,
        artist TEXT NOT NULL,
        album TEXT NOT NULL,
        year INTEGER,
        trackNumber INTEGER,
        discNumber INTEGER,
        originalTrackNumber INTEGER,
        genre TEXT NOT NULL,
        duration REAL NOT NULL,
        fileSize INTEGER NOT NULL,
        albumArtUrl TEXT,
        albumArtThumbUrl TEXT,
        storageUrl TEXT NOT NULL,
        storagePath TEXT NOT NULL,
        fileName TEXT NOT NULL,
        fileType TEXT NOT NULL,
        dateAdded INTEGER NOT NULL,
        createdAt INTEGER NOT NULL,
        updatedAt INTEGER NOT NULL
      )
    `);

    // Create index for library queries
    this.db.run(`
      CREATE INDEX IF NOT EXISTS idx_songs_library ON songs(libraryId)
    `);

    // Notes table
    this.db.run(`
      CREATE TABLE IF NOT EXISTS notes (
        id TEXT PRIMARY KEY,
        type TEXT NOT NULL,
        title TEXT NOT NULL,
        parentId TEXT,
        content TEXT,
        date TEXT,
        time TEXT,
        tags TEXT,
        images TEXT,
        gridSize TEXT,
        sortMode TEXT,
        musicSortColumn TEXT,
        musicSortDirection TEXT,
        published INTEGER,
        slug TEXT,
        sortOrder INTEGER,
        createdAt INTEGER NOT NULL,
        updatedAt INTEGER NOT NULL
      )
    `);

    // Create index for parent queries
    this.db.run(`
      CREATE INDEX IF NOT EXISTS idx_notes_parent ON notes(parentId)
    `);

    console.log("[LocalCache] Tables created");
  }

  /**
   * Persist database to IndexedDB
   */
  async persist(): Promise<void> {
    const data = this.db.export();
    await saveToIndexedDB(data);
    console.log("[LocalCache] Persisted to IndexedDB");
  }

  // ============================================
  // SONGS METHODS
  // ============================================

  /**
   * Get all songs for a library (instant local read)
   */
  getSongsByLibrary(libraryId: string): Song[] {
    const stmt = this.db.prepare(
      "SELECT * FROM songs WHERE libraryId = ?"
    );
    stmt.bind([libraryId]);

    const songs: Song[] = [];
    while (stmt.step()) {
      songs.push(this.rowToSong(stmt.getAsObject()));
    }
    stmt.free();
    return songs;
  }

  /**
   * Get a single song by ID
   */
  getSongById(id: string): Song | null {
    const stmt = this.db.prepare("SELECT * FROM songs WHERE id = ?");
    stmt.bind([id]);

    if (stmt.step()) {
      const song = this.rowToSong(stmt.getAsObject());
      stmt.free();
      return song;
    }
    stmt.free();
    return null;
  }

  /**
   * Upsert a song (insert or replace)
   */
  upsertSong(song: Song): void {
    if (!song.id) return;

    this.db.run(`
      INSERT OR REPLACE INTO songs (
        id, libraryId, title, artist, album, year, trackNumber,
        discNumber, originalTrackNumber, genre, duration, fileSize,
        albumArtUrl, albumArtThumbUrl, storageUrl, storagePath,
        fileName, fileType, dateAdded, createdAt, updatedAt
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `, [
      song.id,
      song.libraryId || "",
      song.title || "",
      song.artist || "",
      song.album || "",
      song.year ?? null,
      song.trackNumber ?? null,
      song.discNumber ?? null,
      song.originalTrackNumber ?? null,
      song.genre || "",
      song.duration || 0,
      song.fileSize || 0,
      song.albumArtUrl ?? null,
      song.albumArtThumbUrl ?? null,
      song.storageUrl || "",
      song.storagePath || "",
      song.fileName || "",
      song.fileType || "",
      timestampToMillis(song.dateAdded),
      timestampToMillis(song.createdAt),
      timestampToMillis(song.updatedAt),
    ]);
  }

  /**
   * Bulk upsert songs (for initial sync)
   */
  upsertSongs(songs: Song[]): void {
    this.db.run("BEGIN TRANSACTION");
    try {
      for (const song of songs) {
        this.upsertSong(song);
      }
      this.db.run("COMMIT");
    } catch (e) {
      this.db.run("ROLLBACK");
      throw e;
    }
  }

  /**
   * Delete a song
   */
  deleteSong(id: string): void {
    this.db.run("DELETE FROM songs WHERE id = ?", [id]);
  }

  /**
   * Delete all songs in a library
   */
  deleteSongsByLibrary(libraryId: string): void {
    this.db.run("DELETE FROM songs WHERE libraryId = ?", [libraryId]);
  }

  /**
   * Convert SQLite row to Song object
   */
  private rowToSong(row: Record<string, unknown>): Song {
    return {
      id: row.id as string,
      libraryId: row.libraryId as string,
      title: row.title as string,
      artist: row.artist as string,
      album: row.album as string,
      year: row.year as number | null,
      trackNumber: row.trackNumber as number | null,
      discNumber: row.discNumber as number | null,
      originalTrackNumber: row.originalTrackNumber as number | null,
      genre: row.genre as string,
      duration: row.duration as number,
      fileSize: row.fileSize as number,
      albumArtUrl: row.albumArtUrl as string | null,
      albumArtThumbUrl: row.albumArtThumbUrl as string | null,
      storageUrl: row.storageUrl as string,
      storagePath: row.storagePath as string,
      fileName: row.fileName as string,
      fileType: row.fileType as string,
      dateAdded: millisToTimestamp(row.dateAdded as number),
      createdAt: millisToTimestamp(row.createdAt as number),
      updatedAt: millisToTimestamp(row.updatedAt as number),
    };
  }

  // ============================================
  // NOTES METHODS
  // ============================================

  /**
   * Get all notes (instant local read)
   */
  getAllNotes(): NoteItem[] {
    const stmt = this.db.prepare("SELECT * FROM notes ORDER BY sortOrder ASC, title COLLATE NOCASE ASC");

    const notes: NoteItem[] = [];
    while (stmt.step()) {
      notes.push(this.rowToNote(stmt.getAsObject()));
    }
    stmt.free();
    return notes;
  }

  /**
   * Get notes by parent ID
   */
  getNotesByParent(parentId: string | null): NoteItem[] {
    const stmt = this.db.prepare(
      parentId === null
        ? "SELECT * FROM notes WHERE parentId IS NULL ORDER BY sortOrder ASC, title COLLATE NOCASE ASC"
        : "SELECT * FROM notes WHERE parentId = ? ORDER BY sortOrder ASC, title COLLATE NOCASE ASC"
    );
    if (parentId !== null) {
      stmt.bind([parentId]);
    }

    const notes: NoteItem[] = [];
    while (stmt.step()) {
      notes.push(this.rowToNote(stmt.getAsObject()));
    }
    stmt.free();
    return notes;
  }

  /**
   * Get a single note by ID
   */
  getNoteById(id: string): NoteItem | null {
    const stmt = this.db.prepare("SELECT * FROM notes WHERE id = ?");
    stmt.bind([id]);

    if (stmt.step()) {
      const note = this.rowToNote(stmt.getAsObject());
      stmt.free();
      return note;
    }
    stmt.free();
    return null;
  }

  /**
   * Upsert a note (insert or replace)
   */
  upsertNote(note: NoteItem): void {
    if (!note.id) return;

    this.db.run(`
      INSERT OR REPLACE INTO notes (
        id, type, title, parentId, content, date, time, tags,
        images, gridSize, sortMode, musicSortColumn, musicSortDirection,
        published, slug, sortOrder, createdAt, updatedAt
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `, [
      note.id,
      note.type || "note",
      note.title || "",
      note.parentId ?? null,
      note.content ?? null,
      note.date ?? null,
      note.time ?? null,
      note.tags ? JSON.stringify(note.tags) : null,
      note.images ? JSON.stringify(note.images) : null,
      note.gridSize ?? null,
      note.sortMode ?? null,
      note.musicSortColumn ?? null,
      note.musicSortDirection ?? null,
      note.published ? 1 : 0,
      note.slug ?? null,
      note.sortOrder ?? null,
      timestampToMillis(note.createdAt),
      timestampToMillis(note.updatedAt),
    ]);
  }

  /**
   * Bulk upsert notes (for initial sync)
   */
  upsertNotes(notes: NoteItem[]): void {
    this.db.run("BEGIN TRANSACTION");
    try {
      for (const note of notes) {
        this.upsertNote(note);
      }
      this.db.run("COMMIT");
    } catch (e) {
      this.db.run("ROLLBACK");
      throw e;
    }
  }

  /**
   * Delete a note
   */
  deleteNote(id: string): void {
    this.db.run("DELETE FROM notes WHERE id = ?", [id]);
  }

  /**
   * Convert SQLite row to NoteItem object
   */
  private rowToNote(row: Record<string, unknown>): NoteItem {
    return {
      id: row.id as string,
      type: row.type as NoteItem["type"],
      title: row.title as string,
      parentId: row.parentId as string | null,
      content: row.content as string | undefined,
      date: row.date as string | undefined,
      time: row.time as string | null | undefined,
      tags: row.tags ? JSON.parse(row.tags as string) : undefined,
      images: row.images ? JSON.parse(row.images as string) : undefined,
      gridSize: row.gridSize as NoteItem["gridSize"],
      sortMode: row.sortMode as NoteItem["sortMode"],
      musicSortColumn: row.musicSortColumn as NoteItem["musicSortColumn"],
      musicSortDirection: row.musicSortDirection as NoteItem["musicSortDirection"],
      published: row.published === 1,
      slug: row.slug as string | undefined,
      sortOrder: row.sortOrder as number | undefined,
      createdAt: millisToTimestamp(row.createdAt as number),
      updatedAt: millisToTimestamp(row.updatedAt as number),
    };
  }

  // ============================================
  // UTILITY METHODS
  // ============================================

  /**
   * Clear all data (for testing/reset)
   */
  clear(): void {
    this.db.run("DELETE FROM songs");
    this.db.run("DELETE FROM notes");
    console.log("[LocalCache] Cleared all data");
  }

  /**
   * Get cache stats
   */
  getStats(): { songs: number; notes: number } {
    const songsResult = this.db.exec("SELECT COUNT(*) FROM songs");
    const notesResult = this.db.exec("SELECT COUNT(*) FROM notes");

    return {
      songs: songsResult[0]?.values[0]?.[0] as number || 0,
      notes: notesResult[0]?.values[0]?.[0] as number || 0,
    };
  }
}

// ============================================
// IndexedDB persistence helpers
// ============================================

const DB_NAME = "dirigible-cache";
const STORE_NAME = "sqlite";
const DB_KEY = "database";

async function openIndexedDB(): Promise<IDBDatabase> {
  return new Promise((resolve, reject) => {
    const request = indexedDB.open(DB_NAME, 1);

    request.onerror = () => reject(request.error);
    request.onsuccess = () => resolve(request.result);

    request.onupgradeneeded = (event) => {
      const db = (event.target as IDBOpenDBRequest).result;
      if (!db.objectStoreNames.contains(STORE_NAME)) {
        db.createObjectStore(STORE_NAME);
      }
    };
  });
}

async function saveToIndexedDB(data: Uint8Array): Promise<void> {
  const db = await openIndexedDB();
  return new Promise((resolve, reject) => {
    const tx = db.transaction(STORE_NAME, "readwrite");
    const store = tx.objectStore(STORE_NAME);
    const request = store.put(data, DB_KEY);

    request.onerror = () => reject(request.error);
    request.onsuccess = () => resolve();
  });
}

async function loadFromIndexedDB(): Promise<Uint8Array | null> {
  const db = await openIndexedDB();
  return new Promise((resolve, reject) => {
    const tx = db.transaction(STORE_NAME, "readonly");
    const store = tx.objectStore(STORE_NAME);
    const request = store.get(DB_KEY);

    request.onerror = () => reject(request.error);
    request.onsuccess = () => resolve(request.result || null);
  });
}

// ============================================
// Timestamp conversion helpers
// ============================================

function timestampToMillis(ts: Timestamp | undefined): number {
  if (!ts) return Date.now();
  if (typeof ts.toMillis === "function") return ts.toMillis();
  // Handle plain object from Firestore
  if (ts && typeof ts === "object" && "seconds" in ts) {
    return (ts as { seconds: number; nanoseconds: number }).seconds * 1000;
  }
  return Date.now();
}

function millisToTimestamp(millis: number): Timestamp {
  return Timestamp.fromMillis(millis);
}
